File: mobile-apps/alloy-ios/Alloy/Denon/DenonConnection.swift
import Foundation
import Network

/// A persistent TCP line-based connection to a Denon AVR. Emits raw
/// response lines as they arrive. Reconnects with exponential backoff.
///
/// The connection carries both our own command replies *and* unsolicited
/// push updates from the device — both look the same on the wire, so
/// consumers just parse everything through `DenonCommand.parse`.
actor DenonConnection {
    private let host: String
    private let port: Int
    private var connection: NWConnection?
    private var readBuffer = Data()
    private var backoff: TimeInterval = 1.0
    private let maxBackoff: TimeInterval = 30.0

    private var lineContinuation: AsyncStream<String>.Continuation?
    private var stateContinuation: AsyncStream<ConnectionState>.Continuation?

    nonisolated let lineStream: AsyncStream<String>
    nonisolated let stateStream: AsyncStream<ConnectionState>

    init(host: String, port: Int) {
        self.host = host
        self.port = port
        var lineCont: AsyncStream<String>.Continuation!
        var stateCont: AsyncStream<ConnectionState>.Continuation!
        self.lineStream = AsyncStream { lineCont = $0 }
        self.stateStream = AsyncStream { stateCont = $0 }
        self.lineContinuation = lineCont
        self.stateContinuation = stateCont
    }

    func start() async {
        stateContinuation?.yield(.connecting)
        let params = NWParameters.tcp
        let endpoint = NWEndpoint.hostPort(
            host: .init(host),
            port: .init(integerLiteral: UInt16(port))
        )
        let conn = NWConnection(to: endpoint, using: params)
        self.connection = conn

        conn.stateUpdateHandler = { [weak self] state in
            Task { await self?.handleState(state) }
        }
        conn.start(queue: .global(qos: .userInitiated))
        await receiveLoop()
    }

    func stop() {
        connection?.cancel()
        connection = nil
        stateContinuation?.yield(.disconnected)
    }

    func send(_ line: String) async {
        guard let data = line.data(using: .ascii), let conn = connection else { return }
        conn.send(content: data, completion: .contentProcessed { _ in })
    }

    // MARK: - Private

    private func handleState(_ state: NWConnection.State) async {
        switch state {
        case .ready:
            backoff = 1.0
            stateContinuation?.yield(.connected)
        case .failed(let error), .waiting(let error):
            stateContinuation?.yield(.failed(error.localizedDescription))
            await scheduleReconnect()
        case .cancelled:
            stateContinuation?.yield(.disconnected)
        default:
            break
        }
    }

    private func scheduleReconnect() async {
        let delay = backoff
        backoff = min(backoff * 2, maxBackoff)
        try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
        await start()
    }

    private func receiveLoop() async {
        guard let conn = connection else { return }
        conn.receive(minimumIncompleteLength: 1, maximumLength: 4096) { [weak self] data, _, isComplete, error in
            Task { [weak self] in
                guard let self else { return }
                if let data, !data.isEmpty {
                    await self.append(data)
                }
                if error != nil || isComplete {
                    await self.scheduleReconnect()
                } else {
                    await self.receiveLoop()
                }
            }
        }
    }

    private func append(_ data: Data) {
        readBuffer.append(data)
        // Denon responses are terminated by \r (0x0D). Split on CR.
        while let idx = readBuffer.firstIndex(of: 0x0D) {
            let lineData = readBuffer[..<idx]
            readBuffer.removeSubrange(...idx)
            if let line = String(data: lineData, encoding: .ascii), !line.isEmpty {
                lineContinuation?.yield(line)
            }
        }
    }
}

File: mobile-apps/alloy-android/app/src/main/kotlin/com/sharca/alloy/denon/DenonConnection.kt
package com.sharca.alloy.denon

import com.sharca.alloy.model.ConnectionState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStream
import java.net.Socket

/**
 * Persistent TCP line-based connection to a Denon AVR. Emits raw
 * response lines as they arrive. Reconnects with exponential backoff.
 *
 * The connection carries both our own command replies *and*
 * unsolicited push updates from the device — both look the same on
 * the wire, so consumers just parse everything through
 * [DenonCommand.parse].
 */
class DenonConnection(
    private val host: String,
    private val port: Int = 23
) {
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val _lines = MutableSharedFlow<String>(extraBufferCapacity = 64)
    val lines: SharedFlow<String> = _lines.asSharedFlow()

    private val _state = MutableStateFlow<ConnectionState>(ConnectionState.Disconnected)
    val state: StateFlow<ConnectionState> = _state.asStateFlow()

    private var socket: Socket? = null
    private var writer: OutputStream? = null
    private var loopJob: Job? = null
    private var backoff: Long = 1_000L
    private val maxBackoff: Long = 30_000L

    fun start() {
        if (loopJob?.isActive == true) return
        loopJob = scope.launch { connectLoop() }
    }

    fun stop() {
        loopJob?.cancel()
        loopJob = null
        try { socket?.close() } catch (_: Throwable) {}
        socket = null
        _state.value = ConnectionState.Disconnected
    }

    suspend fun send(line: String) = withContext(Dispatchers.IO) {
        try {
            writer?.write(line.toByteArray(Charsets.US_ASCII))
            writer?.flush()
        } catch (e: Throwable) {
            _state.value = ConnectionState.Failed(e.message ?: "send failed")
        }
    }

    // MARK: - Private

    private suspend fun connectLoop() {
        while (scope.isActive) {
            _state.value = ConnectionState.Connecting
            try {
                val s = Socket(host, port)
                socket = s
                writer = s.getOutputStream()
                _state.value = ConnectionState.Connected
                backoff = 1_000L

                val reader = BufferedReader(InputStreamReader(s.getInputStream(), Charsets.US_ASCII))
                val buffer = StringBuilder()
                val chars = CharArray(256)
                while (scope.isActive) {
                    val n = reader.read(chars)
                    if (n < 0) break
                    for (i in 0 until n) {
                        val ch = chars[i]
                        if (ch == '\r') {
                            if (buffer.isNotEmpty()) {
                                _lines.emit(buffer.toString())
                                buffer.clear()
                            }
                        } else if (ch != '\n') {
                            buffer.append(ch)
                        }
                    }
                }
                _state.value = ConnectionState.Disconnected
            } catch (e: Throwable) {
                _state.value = ConnectionState.Failed(e.message ?: "connect failed")
            } finally {
                try { socket?.close() } catch (_: Throwable) {}
                socket = null
                writer = null
            }
            if (!scope.isActive) break
            delay(backoff)
            backoff = (backoff * 2).coerceAtMost(maxBackoff)
        }
    }
}
